2020-01-Winter/2_string_manipulation/README.md

```diff
@@ -116,21 +133,66 @@ Go to [Top](#top)
 <!-- Don't remove -->
 <a name="s2"/>
 
-### 2. SOLUTION 2 TODO :bug:
+### 2. GROUP ANAGRAMS
 
-Source: TODO :bug:
+Two strings are anagrams if their sorted strings are equal. Then, one way
+to solve this problem is by categorizing by sorting arrays.
 
-#### Naive/Simple Solution
+1. Check for an empty array, if the array is not empty, then use a hashtable
+where the key is the sorted string, and the value is the string that once is sorted, 
+it matches the key.
 
-TODO :bug:
+2. Iterate through the array of strings. For each string in the array of strings,
+create an array of chars, using the .toCharArray() function. 
+
+3. Sort the char array, and convert it into a string. This string will be used as the
+key for the key-value pair of the hashtable.
+
+4. If the hashtable does not contains the key(string value of sorted char array), add it to the map. 
+
+5. Add the original (unsorted) string as a new value of the key.
+ 
+6. Return the list of list of anagrams
+
+Time Complexity: The outer loop has complexity O(N) as we iterate through each string. 
+Then, we sort each string in O(N log N) time. Therefore, the time complexity of this 
+algorithm is O(N Log N)
+
+Space Complexity: O(N)
 
```

Comment:
Missing the solution code here; add it in like so:

````suggestion

```java
public static List<List<String>> groupAnagramsOpt(String[] strs) {
    if (strs.length == 0) {
        return new ArrayList<>();
    }
    Map<String, List<String>> map = new HashMap<>();    
    for (String st:strs) {
        char[] chars = st.toCharArray();    
        Arrays.sort(chars);    
        String key = String.valueOf(chars); // Converting the sorted char sequence into string    
        if (!map.containsKey(key)) {
            map.put(key, new ArrayList<String>());
        }    
        map.get(key).add(st);
    }
    return new ArrayList(map.values());
}
```

````